### PR TITLE
#9466: handling GetFeatureInfo exceptions parameter format configuration

### PR DESCRIPTION
**Pull request by mahmoudadel54**
## Description
- Handling GFI exceptions parameter format by removing "application/json" and putting the geoserver default format for exceptions 'application/vnd.ogc.se_xml' that will work with all WMS servcies including arcgis services.
- Add getIdentifyWorkflow to catch the GFI errors and exceptions. 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix

## Issue
#9466 

**What is the current behavior?**
Some services has no "application/json" into their exception format list. So an exception of "Parameter 'exceptions' contains unacceptable value" shown for the GFI and prevent showing the identify results.

#<issue>
#9466 

**What is the new behavior?**
There is no exception related to wrong exception format in GFI and identify works well with all WMS servcies including arcgis services and the identify results shown to user normally.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] No

### CODE REVIEW

**Review by tdipisa — COMMENTED** [review]
@mahmoudadel54 it seems the unit test check is failing. Can you please check?

File: web/client/utils/mapinfo/wms.js [hunk 1]
```diff
@@ -59,7 +62,7 @@ export default {
                 service: 'WMS',
                 version: '1.1.1',
                 request: 'GetFeatureInfo',
-                exceptions: 'application/json',
+                exceptions: 'application/vnd.ogc.se_xml',           // the default exception format
```
Comment (offtherailz) [inline web/client/utils/mapinfo/wms.js:65]: I think we can remove this parameter at all... not sure to force it.

**Review by offtherailz — CHANGES_REQUESTED** [review]
Please:
-  `getIdentifyFlow` should have a unit test that checks the effective output for exception.
- the failing unit tests must be fixed

**Review by offtherailz — APPROVED** [review]
I effectively see that no event is triggered by openlayers for these load events.
So there is not an easy solution.
Even if we can trigger the load start on add, we can not catch load end... so I think it could be better to open a dedicated issue for that.
@tdipisa let me know what do you think. The rest is ready to be merged.

**Comment by offtherailz** [comment]
sorry I sent my review to this PR in error, but the comment was for another pull request. 

:facepalm:

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2022, GeoSolutions Sas.
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:2]: this is a minor thing, but for new files the copyright has to be aligned to current year
```suggestion
 * Copyright 2023, GeoSolutions Sas.
```

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import expect from 'expect';
+
+import axios from '../../../libs/ajax';
+import MockAdapter from "axios-mock-adapter";
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:12]: when you can try to keep external libraries separated by project modules and to be listed as the first thing
```suggestion
import expect from 'expect';
import MockAdapter from "axios-mock-adapter";

import axios from '../../../libs/ajax';
```

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import {INFO_FORMATS} from "../../FeatureInfoUtils";
+import {getFeatureInfo} from "../../../api/identify";
+
+describe('mapinfo wms utils', () => {
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:16]: can you add some test that make intercept ogc error to be triggered?

**Review by MV88 — CHANGES_REQUESTED** [review]
a few things
the rest lgtm

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 5]
```diff
@@ -108,4 +110,38 @@ describe('mapinfo wms utils', () => {
...
+    it('test intercept ogc error in wms if success', (done)=>{
+        mockAxios.onGet().reply(() => {
+            return [200, {
+                "type": "FeatureCollection",
+                "features": [],
+                "totalFeatures": "unknown",
+                "numberReturned": 0,
+                "timeStamp": "2023-09-22T08:50:30.808Z",
+                "crs": null
+            }];
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:122]: I think here you can return the body of an exception in json

try to get it by requesting  a get feature info with exception format in application/json  for a layer that does not exist maybe

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 6]
```diff
@@ -108,4 +110,38 @@ describe('mapinfo wms utils', () => {
...
+        });
+        wms
+            .getIdentifyFlow(undefined, "/", { features: [] })
+            .subscribe((response) => {
+                expect(response?.data?.features).toEqual([]);
+                done();
+            }, error => {
+                done(error);
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:130]: just call done()

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 7]
```diff
@@ -108,4 +110,38 @@ describe('mapinfo wms utils', () => {
...
+                "numberReturned": 0,
+                "timeStamp": "2023-09-22T08:50:30.808Z",
+                "crs": null
+            }];
+        });
+        wms
+            .getIdentifyFlow(undefined, "/", { features: [] })
+            .subscribe((response) => {
+                expect(response?.data?.features).toEqual([]);
+                done();
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:128]: remove this call so that we are sure it works because it triggers the error call back

**Review by MV88 — CHANGES_REQUESTED** [review]

File: web/client/utils/mapinfo/__tests__/wms-test.js (hunk 6 above)
Comment (mahmoudadel54) [reply web/client/utils/mapinfo/__tests__/wms-test.js]: It gave eslint error here due to not using error

Comment (MV88) [reply web/client/utils/mapinfo/__tests__/wms-test.js]: then remove the error param

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                        "text": "Could not find layer misc:Bonnebladen1900_index2121 Details:org.geoserver.platform.ServiceException: Could not find layer misc:Bonnebladen1900_index2121"
+                    }]
+            }];
+        });
+        wms
+            .getIdentifyFlow(undefined, "/", { features: [] })
+            .subscribe((response) => {
+                expect(response?.data?.exceptions).toExist();
+                done();
+            });
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:148]: are you sure this is actually testing the intercept ogc error call?

shouldn't throw an error and therefore the error callback should be called. so done have to placed there

File: web/client/utils/mapinfo/__tests__/wms-test.js [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        mockAxios.onGet().reply(() => {
+            return [404, {}];
+        });
+        wms
+            .getIdentifyFlow(undefined, "/", { features: [] })
+            .subscribe((response) => {
+                expect(response?.data?.features).toEqual([]);
+            }, error => {
+                expect(error.status).toEqual(404);
+                done();
```
Comment (MV88) [inline web/client/utils/mapinfo/__tests__/wms-test.js:160]: see previous comment, it should be similar to this one

**Comment by ElenaGallo** [comment]
Test passe, @mahmoudadel54 please backport this to 2023.02.xx. Thanks

**Pull request merged**